pkg/client/userd/trafficmgr/gather_logs.go

```diff
@@ -0,0 +1,156 @@
+	req := podsAPI.GetLogs(pod.Name, &core.PodLogOptions{Container: container})
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get log for %s.%s: %v", pod.Name, pod.Namespace, err)
+		dlog.Error(ctx, msg)
+		return msg, ""
+	}
+	defer podLogs.Close()
+
+	buf := new(bytes.Buffer)
+	if _, err = io.Copy(buf, podLogs); err != nil {
+		msg := fmt.Sprintf("Failed writing log to buffer: %v", err)
+		dlog.Error(ctx, msg)
+		return msg, ""
+	}
+	podLog := buf.String()
+
+	// Get the pod yaml if the user asked for it
+	if podYAML {
+		var b []byte
+		if b, err = yaml.Marshal(pod); err != nil {
+			msg := fmt.Sprintf("Failed marshaling pod yaml: %v", err)
+			dlog.Error(ctx, msg)
+			return msg, ""
+		}
+		return podLog, string(b)
+	}
+	return podLog, ""
+}
+
+// GatherLogs acquires the logs for the traffic-manager and/or traffic-agents specified by the
+// connector.LogsRequest and returns them to the caller
+func (tm *TrafficManager) GatherLogs(ctx context.Context, request *connector.LogsRequest) (*connector.LogsResponse, error) {
+	logWriteMutex := &sync.Mutex{}
+	coreAPI := k8sapi.GetK8sInterface(ctx).CoreV1()
+	resp := &connector.LogsResponse{
+		PodLogs: make(map[string]string),
+		PodYaml: make(map[string]string),
+	}
+	container := "traffic-agent"
+	hasContainer := func(pod *core.Pod) bool {
+		if strings.EqualFold(request.Agents, "all") || strings.Contains(pod.Name, request.Agents) {
+			cns := pod.Spec.Containers
+			for c := range cns {
+				if cns[c].Name == container {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	if !strings.EqualFold(request.Agents, "none") {
+		for _, ns := range tm.GetCurrentNamespaces(true) {
+			podsAPI := coreAPI.Pods(ns)
+			podList, err := podsAPI.List(ctx, meta.ListOptions{})
+			if err != nil {
+				resp.ErrMsg = err.Error()
+				dlog.Error(ctx, resp.ErrMsg)
+				return resp, nil
+			}
+			pods := podList.Items
+			podsWithContainer := make([]*core.Pod, 0, len(pods))
+			for i := range pods {
+				pod := &pods[i]
+				if hasContainer(pod) {
+					podsWithContainer = append(podsWithContainer, pod)
+				}
+			}
+			wg := sync.WaitGroup{}
+			wg.Add(len(podsWithContainer))
+			for _, pod := range podsWithContainer {
+				go func(pod *core.Pod) {
+					defer wg.Done()
+					// Since the same named workload could exist in multiple namespaces
+					// we add the namespace into the name so that it's easier to make
+					// sense of the logs
+					podAndNs := fmt.Sprintf("%s.%s", pod.Name, pod.Namespace)
+					dlog.Debugf(ctx, "gathering logs for %s, yaml = %t", podAndNs, request.GetPodYaml)
+					logs, yml := getPodLog(ctx, podsAPI, pod, container, request.GetPodYaml)
+					logWriteMutex.Lock()
+					resp.PodLogs[podAndNs] = logs
+					if request.GetPodYaml {
+						resp.PodYaml[podAndNs] = yml
+					}
+					logWriteMutex.Unlock()
+				}(pod)
+			}
+			wg.Wait()
```

Comment:
could we do all ns in parallel instead of one ns at a time?

Reply:
This solution will do all pods in a namespace in parallel. Extended to do all namespaces in parallel too, means all pods with agents in parallel. I'm concerned that it may result in too many simultaneous connections. Given that the command isn't performance critical, I think I'd like to keep it as-is.